File: sdks/go/cmd/elydora/plugins/cursor_contract.go
package plugins

import (
	"encoding/json"
	"fmt"
	"path/filepath"
	"reflect"
	"runtime"
	"sort"
	"strings"
)

const (
	cursorAgentKey    = "cursor"
	cursorConfigFile  = "hooks.json"
	cursorGuardScript = "guard.js"
	cursorAuditScript = "hook.js"
	cursorHookTimeout = float64(10)
)

type cursorHooks map[string][]map[string]any

type cursorDocument struct {
	exists   bool
	filePath string
	root     map[string]any
	hooks    cursorHooks
	raw      []byte
}

type cursorRenderedDocument struct {
	document *cursorDocument
	changed  bool
	next     []byte
	remove   bool
}

type cursorRuntimeContract = managedRuntimeContract

func cloneCursorHooks(source cursorHooks) cursorHooks {
	clone := make(cursorHooks, len(source))
	for event, handlers := range source {
		clone[event] = append([]map[string]any(nil), handlers...)
	}
	return clone
}

func buildCursorHandler(nodePath, scriptPath string) map[string]any {
	command := posixSource(nodePath, scriptPath)
	if runtime.GOOS == "windows" {
		command = powerShellSource(nodePath, scriptPath)
	}
	return map[string]any{
		"command": command, "timeout": cursorHookTimeout, "failClosed": true,
	}
}

func parseCursorLegacyCommand(command string) (string, bool) {
	separator := strings.IndexByte(command, ' ')
	if separator < 0 {
		return "", false
	}
	executable := command[:separator]
	if !strings.EqualFold(executable, "node") && !strings.EqualFold(executable, "node.exe") {
		return "", false
	}
	script := command[separator+1:]
	if len(script) >= 2 && script[0] == '"' && script[len(script)-1] == '"' {
		script = script[1 : len(script)-1]
	}
	return script, script != "" && !strings.ContainsAny(script, "\r\n")
}

func cursorManagedScriptPath(handler map[string]any) (string, bool) {
	if len(handler) == 3 && handler["timeout"] == cursorHookTimeout && handler["failClosed"] == true {
		command, ok := handler["command"].(string)
		if !ok {
			return "", false
		}
		var nodePath, scriptPath string
		var parsed bool
		if runtime.GOOS == "windows" {
			nodePath, scriptPath, parsed = parsePowerShellSource(command)
		} else {
			nodePath, scriptPath, parsed = parsePOSIXCommand(command)
		}
		if parsed && filepath.IsAbs(nodePath) && filepath.IsAbs(scriptPath) &&
			isNodeExecutable(nodePath) {
			return scriptPath, true
		}
	}
	if len(handler) != 1 {
		return "", false
	}
	command, ok := handler["command"].(string)
	if !ok {
		return "", false
	}
	return parseCursorLegacyCommand(command)
}

func cursorManagedAgentID(
	handler map[string]any,
	scriptName string,
	runtimeRoot string,
) (string, bool) {
	scriptPath, managed := cursorManagedScriptPath(handler)
	if !managed || !filepath.IsAbs(scriptPath) || filepath.Base(scriptPath) != scriptName {
		return "", false
	}
	agentDirectory := filepath.Dir(scriptPath)
	if !sameManagedPath(filepath.Dir(agentDirectory), runtimeRoot) {
		return "", false
	}
	agentID := filepath.Base(agentDirectory)
	return agentID, agentID != "" && agentID != "." && agentID != ".."
}

func readCursorHooks(value any, label string) (cursorHooks, error) {
	object, ok := value.(map[string]any)
	if !ok || object == nil {
		return nil, fmt.Errorf(`%s field "hooks" must be an object`, label)
	}
	hooks := make(cursorHooks, len(object))
	for event, handlerValue := range object {
		values, ok := handlerValue.([]any)
		if !ok {
			return nil, fmt.Errorf(`%s field "hooks.%s" must be an array`, label, event)
		}
		handlers := make([]map[string]any, 0, len(values))
		for index, value := range values {
			handler, ok := value.(map[string]any)
			if !ok || handler == nil {
				return nil, fmt.Errorf(`%s handler hooks.%s[%d] must be an object`, label, event, index)
			}
			handlers = append(handlers, handler)
		}
		hooks[event] = handlers
	}
	return hooks, nil
}

func cursorContainsManagedHook(hooks cursorHooks, runtimeRoot string) bool {
	for _, contract := range []struct{ event, script string }{
		{"preToolUse", cursorGuardScript},
		{"postToolUse", cursorAuditScript},
		{"postToolUseFailure", cursorAuditScript},
	} {
		for _, handler := range hooks[contract.event] {
			if _, managed := cursorManagedAgentID(handler, contract.script, runtimeRoot); managed {
				return true
			}
		}
	}
	return false
}

func parseCursorDocument(filePath string, raw []byte, runtimeRoot string) (*cursorDocument, error) {
	label := fmt.Sprintf("Cursor user hooks at %s", filePath)
	if !json.Valid(raw) {
		var value any
		if err := json.Unmarshal(raw, &value); err != nil {
			return nil, fmt.Errorf("parse %s: %w", label, err)
		}
	}
	root, err := decodeJSONCObject(raw, label, false)
	if err != nil {
		return nil, err
	}
	hooks := cursorHooks{}
	if value, exists := root["hooks"]; exists {
		hooks, err = readCursorHooks(value, label)
		if err != nil {
			return nil, err
		}
	}
	version, hasVersion := root["version"]
	current := hasVersion && version == float64(1)
	if !current && (hasVersion || !cursorContainsManagedHook(hooks, runtimeRoot)) {
		return nil, fmt.Errorf("%s must declare version 1", label)
	}
	return &cursorDocument{
		exists: true, filePath: filePath, root: root,
		hooks: hooks, raw: append([]byte(nil), raw...),
	}, nil
}

func createCursorDocument(filePath string) *cursorDocument {
	return &cursorDocument{filePath: filePath, root: map[string]any{}, hooks: cursorHooks{}}
}

func removeManagedCursorHooks(hooks cursorHooks, runtimeRoot, agentID string) cursorHooks {
	result := cloneCursorHooks(hooks)
	for _, contract := range []struct{ event, script string }{
		{"preToolUse", cursorGuardScript},
		{"postToolUse", cursorAuditScript},
		{"postToolUseFailure", cursorAuditScript},
	} {
		kept := make([]map[string]any, 0, len(result[contract.event]))
		for _, handler := range result[contract.event] {
			managedID, managed := cursorManagedAgentID(handler, contract.script, runtimeRoot)
			if managed && (agentID == "" || sameManagedAgentID(managedID, agentID)) {
				continue
			}
			kept = append(kept, handler)
		}
		if len(kept) == 0 {
			delete(result, contract.event)
		} else {
			result[contract.event] = kept
		}
	}
	return result
}

func entirelyManagedCursorDocument(document *cursorDocument, runtimeRoot string) bool {
	if !document.exists || len(document.hooks) == 0 {
		return false
	}
	for key := range document.root {
		if key != "version" && key != "hooks" {
			return false
		}
	}
	count := 0
	for event, handlers := range document.hooks {
		script := map[string]string{
			"preToolUse":         cursorGuardScript,
			"postToolUse":        cursorAuditScript,
			"postToolUseFailure": cursorAuditScript,
		}[event]
		if script == "" || len(handlers) == 0 {
			return false
		}
		count += len(handlers)
		for _, handler := range handlers {
			if _, managed := cursorManagedAgentID(handler, script, runtimeRoot); !managed {
				return false
			}
		}
	}
	return count > 0
}

func renderCursorDocument(
	document *cursorDocument,
	hooks cursorHooks,
	runtimeRoot string,
) (*cursorRenderedDocument, error) {
	if !document.exists && len(hooks) == 0 {
		return &cursorRenderedDocument{document: document}, nil
	}
	if document.exists && reflect.DeepEqual(hooks, document.hooks) {
		return &cursorRenderedDocument{document: document}, nil
	}
	if len(hooks) == 0 && entirelyManagedCursorDocument(document, runtimeRoot) {
		return &cursorRenderedDocument{document: document, changed: true, remove: true}, nil
	}
	root := make(map[string]any, len(document.root)+1)
	for key, value := range document.root {
		root[key] = value
	}
	root["version"] = float64(1)
	if len(hooks) == 0 {
		delete(root, "hooks")
	} else {
		root["hooks"] = hooks
	}
	next, err := json.MarshalIndent(root, "", "  ")
	if err != nil {
		return nil, err
	}
	next = append(next, '\n')
	return &cursorRenderedDocument{
		document: document, changed: !document.exists || string(next) != string(document.raw), next: next,
	}, nil
}

func managedCursorIDs(
	handlers []map[string]any,
	scriptName, runtimeRoot string,
) map[string]string {
	result := map[string]string{}
	for _, handler := range handlers {
		agentID, managed := cursorManagedAgentID(handler, scriptName, runtimeRoot)
		if !managed {
			continue
		}
		result[managedReferenceKey(agentID)] = agentID
	}
	return result
}

func cursorRuntimeContracts(hooks cursorHooks, runtimeRoot string) []cursorRuntimeContract {
	guards := managedCursorIDs(hooks["preToolUse"], cursorGuardScript, runtimeRoot)
	audits := managedCursorIDs(hooks["postToolUse"], cursorAuditScript, runtimeRoot)
	failures := managedCursorIDs(hooks["postToolUseFailure"], cursorAuditScript, runtimeRoot)
	keys := make([]string, 0, len(guards))
	for key := range guards {
		_, hasAudit := audits[key]
		_, hasFailureAudit := failures[key]
		if hasAudit && hasFailureAudit {
			keys = append(keys, key)
		}
	}
	sort.Strings(keys)
	contracts := make([]cursorRuntimeContract, 0, len(keys))
	for _, key := range keys {
		agentID := guards[key]
		directory := filepath.Join(runtimeRoot, agentID)
		contracts = append(contracts, cursorRuntimeContract{
			agentID: agentID, guardPath: filepath.Join(directory, cursorGuardScript),
			auditPath: filepath.Join(directory, cursorAuditScript),
		})
	}
	return contracts
}

File: sdks/node/src/plugins/cursor-contract.ts
import os from 'node:os';
import path from 'node:path';
import { sameAgentId, samePath } from './common.js';
import type { GuardScriptOptions } from './guard-template.js';
import type { HookScriptOptions } from './hook-template.js';
import {
  isNodeExecutable,
  parsePosixCommand,
  parsePowerShellSource,
  posixSource,
  powerShellSource,
} from './shell-command.js';
import { isObject, parseStrictJsonObject, type JsonObject } from './strict-json.js';

export const AGENT_KEY = 'cursor';
export const CONFIG_FILE = 'hooks.json';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const HOOK_TIMEOUT_SECONDS = 10;
export const GUARD_OPTIONS: GuardScriptOptions = {
  failClosed: true,
  successOutput: '{"permission":"allow"}\n',
  denyProtocol: 'cursor',
};
export const AUDIT_OPTIONS: HookScriptOptions = {
  failClosed: true,
  nativePayload: true,
  successOutput: '{}\n',
};

export type CursorHooks = Record<string, JsonObject[]>;

export interface CursorDocument {
  readonly exists: boolean;
  readonly filePath: string;
  readonly root: JsonObject;
  readonly hooks: CursorHooks;
  readonly raw?: string;
}

export interface RenderedDocument {
  readonly document: CursorDocument;
  readonly changed: boolean;
  readonly next?: string;
}

export interface RuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
}

export function buildHandler(scriptPath: string): JsonObject {
  const command = process.platform === 'win32'
    ? powerShellSource(scriptPath)
    : posixSource(scriptPath);
  return {
    command,
    timeout: HOOK_TIMEOUT_SECONDS,
    failClosed: true,
  };
}

function parseLegacyCommand(command: string): readonly [string, string] | undefined {
  const match = /^(node(?:\.exe)?)\s+"([^"\r\n]+)"$/i.exec(command);
  return match ? [match[1], match[2]] : undefined;
}

function managedScriptPath(handler: JsonObject): string | undefined {
  const keys = Object.keys(handler);
  if (keys.length === 3
    && typeof handler.command === 'string'
    && handler.timeout === HOOK_TIMEOUT_SECONDS
    && handler.failClosed === true) {
    const parsed = process.platform === 'win32'
      ? parsePowerShellSource(handler.command)
      : parsePosixCommand(handler.command);
    if (parsed
      && path.isAbsolute(parsed[0])
      && path.isAbsolute(parsed[1])
      && isNodeExecutable(parsed[0])) return parsed[1];
  }
  if (keys.length !== 1 || typeof handler.command !== 'string') return undefined;
  const legacy = parseLegacyCommand(handler.command);
  return legacy && isNodeExecutable(legacy[0]) && path.isAbsolute(legacy[1])
    ? legacy[1]
    : undefined;
}

function managedAgentId(handler: JsonObject, scriptName: string): string | undefined {
  const scriptPath = managedScriptPath(handler);
  if (!scriptPath || path.basename(scriptPath) !== scriptName) return undefined;
  const agentDirectory = path.dirname(scriptPath);
  if (!samePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) {
    return undefined;
  }
  const agentId = path.basename(agentDirectory);
  return agentId && agentId !== '.' && agentId !== '..' ? agentId : undefined;
}

function readHooks(value: unknown, label: string): CursorHooks {
  if (value === undefined) return {};
  if (!isObject(value)) throw new Error(`${label} field "hooks" must be an object`);
  const hooks: CursorHooks = {};
  for (const [event, candidate] of Object.entries(value)) {
    if (!Array.isArray(candidate)) {
      throw new Error(`${label} field "hooks.${event}" must be an array`);
    }
    hooks[event] = candidate.map((handler, index) => {
      if (!isObject(handler)) {
        throw new Error(`${label} handler hooks.${event}[${index}] must be an object`);
      }
      return handler;
    });
  }
  return hooks;
}

export function parseDocument(filePath: string, raw: string): CursorDocument {
  const label = `Cursor user hooks at ${filePath}`;
  const root = parseStrictJsonObject(raw, label);
  const hooks = readHooks(root.hooks, label);
  const hasVersion = Object.prototype.hasOwnProperty.call(root, 'version');
  if (root.version !== 1 && (hasVersion || !containsManagedHook(hooks))) {
    throw new Error(`${label} must declare version 1`);
  }
  return {
    exists: true,
    filePath,
    root,
    hooks,
    raw,
  };
}

export function createDocument(filePath: string): CursorDocument {
  return { exists: false, filePath, root: {}, hooks: {} };
}

export function removeManagedHooks(hooks: CursorHooks, agentId?: string): CursorHooks {
  const next: CursorHooks = Object.fromEntries(
    Object.entries(hooks).map(([event, handlers]) => [event, [...handlers]]),
  );
  for (const [event, scriptName] of [
    ['preToolUse', GUARD_SCRIPT],
    ['postToolUse', AUDIT_SCRIPT],
    ['postToolUseFailure', AUDIT_SCRIPT],
  ] as const) {
    const handlers = (next[event] ?? []).filter((handler) => {
      const managedId = managedAgentId(handler, scriptName);
      return !managedId || (agentId !== undefined && !sameAgentId(managedId, agentId));
    });
    if (handlers.length > 0) next[event] = handlers;
    else delete next[event];
  }
  return next;
}

function entirelyManaged(document: CursorDocument): boolean {
  if (!document.exists
    || !Object.keys(document.root).every((key) => key === 'version' || key === 'hooks')) {
    return false;
  }
  const events = Object.entries(document.hooks);
  if (events.length === 0) return false;
  let handlerCount = 0;
  for (const [event, handlers] of events) {
    const scriptName = event === 'preToolUse'
      ? GUARD_SCRIPT
      : event === 'postToolUse' || event === 'postToolUseFailure'
        ? AUDIT_SCRIPT
        : undefined;
    if (!scriptName || handlers.length === 0) return false;
    handlerCount += handlers.length;
    if (handlers.some((handler) => !managedAgentId(handler, scriptName))) return false;
  }
  return handlerCount > 0;
}

export function renderDocument(
  document: CursorDocument,
  hooks: CursorHooks,
): RenderedDocument {
  if (!document.exists && Object.keys(hooks).length === 0) {
    return { document, changed: false };
  }
  if (Object.keys(hooks).length === 0 && entirelyManaged(document)) {
    return { document, changed: true };
  }
  const root: JsonObject = { ...document.root, version: 1 };
  if (Object.keys(hooks).length > 0) root.hooks = hooks;
  else delete root.hooks;
  const next = `${JSON.stringify(root, null, 2)}\n`;
  return { document, changed: next !== document.raw, next };
}

function managedIds(handlers: JsonObject[], scriptName: string): Map<string, string> {
  const result = new Map<string, string>();
  for (const handler of handlers) {
    const agentId = managedAgentId(handler, scriptName);
    if (!agentId) continue;
    const key = process.platform === 'win32' ? agentId.toLowerCase() : agentId;
    result.set(key, agentId);
  }
  return result;
}

function containsManagedHook(hooks: CursorHooks): boolean {
  return [
    ...(hooks.preToolUse ?? []).map((handler) => managedAgentId(handler, GUARD_SCRIPT)),
    ...(hooks.postToolUse ?? []).map((handler) => managedAgentId(handler, AUDIT_SCRIPT)),
    ...(hooks.postToolUseFailure ?? []).map((handler) => managedAgentId(handler, AUDIT_SCRIPT)),
  ].some((agentId) => agentId !== undefined);
}

export function runtimeContracts(hooks: CursorHooks): RuntimeContract[] {
  const guards = managedIds(hooks.preToolUse ?? [], GUARD_SCRIPT);
  const audits = managedIds(hooks.postToolUse ?? [], AUDIT_SCRIPT);
  const failures = managedIds(hooks.postToolUseFailure ?? [], AUDIT_SCRIPT);
  const root = path.join(os.homedir(), '.elydora');
  return [...guards]
    .filter(([key]) => audits.has(key) && failures.has(key))
    .map(([, agentId]) => ({
      agentId,
      guardPath: path.join(root, agentId, GUARD_SCRIPT),
      auditPath: path.join(root, agentId, AUDIT_SCRIPT),
    }));
}
